pkg/libartifact/store/store.go

```diff
@@ -0,0 +1,366 @@
+	if err != nil {
+		return err
+	}
+	_, err = copyer.Copy(ctx, srcRef, destRef)
+	if err != nil {
+		return err
+	}
+	return copyer.Close()
+}
+
+// Push an artifact to an image registry
+func (as ArtifactStore) Push(ctx context.Context, src, dest string, opts libimage.CopyOptions) error {
+	if len(dest) == 0 {
+		return ErrEmptyArtifactName
+	}
+	destRef, err := alltransports.ParseImageName(fmt.Sprintf("docker://%s", dest))
+	if err != nil {
+		return err
+	}
+	srcRef, err := layout.NewReference(as.storePath, src)
+	if err != nil {
+		return err
+	}
+	copyer, err := libimage.NewCopier(&opts, as.SystemContext, nil)
+	if err != nil {
+		return err
+	}
+	_, err = copyer.Copy(ctx, srcRef, destRef)
+	if err != nil {
+		return err
+	}
+	return copyer.Close()
+}
+
+// Add takes one or more local files and adds them to the local artifact store.  The empty
+// string input is for possible custom artifact types.
+func (as ArtifactStore) Add(ctx context.Context, dest string, paths []string, _ string) (*digest.Digest, error) {
+	if len(dest) == 0 {
+		return nil, ErrEmptyArtifactName
+	}
+
+	artifactManifestLayers := make([]specV1.Descriptor, 0)
+
+	// Check if artifact already exists
+	artifacts, err := as.getArtifacts(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check if artifact exists; in GetByName not getting an
+	// error means it exists
+	if _, _, err := artifacts.GetByNameOrDigest(dest); err == nil {
+		return nil, fmt.Errorf("artifact %s already exists", dest)
+	}
+
+	ir, err := layout.NewReference(as.storePath, dest)
+	if err != nil {
+		return nil, err
+	}
+
+	imageDest, err := ir.NewImageDestination(ctx, as.SystemContext)
+	if err != nil {
+		return nil, err
+	}
+	defer imageDest.Close()
+
+	for _, path := range paths {
+		// get the new artifact into the local store
+		newBlobDigest, newBlobSize, err := layout.PutBlobFromLocalFile(ctx, imageDest, path)
+		if err != nil {
+			return nil, err
+		}
+		detectedType, err := determineManifestType(path)
+		if err != nil {
+			return nil, err
+		}
+		newArtifactAnnotations := map[string]string{}
+		newArtifactAnnotations[specV1.AnnotationTitle] = filepath.Base(path)
+		newLayer := specV1.Descriptor{
+			MediaType:   detectedType,
+			Digest:      newBlobDigest,
+			Size:        newBlobSize,
+			Annotations: newArtifactAnnotations,
+		}
+		artifactManifestLayers = append(artifactManifestLayers, newLayer)
+	}
+
+	artifactManifest := specV1.Manifest{
+		Versioned: specs.Versioned{SchemaVersion: 2},
+		MediaType: specV1.MediaTypeImageManifest,
+		// TODO This should probably be configurable once the CLI is capable
+		ArtifactType: "",
+		Config:       specV1.DescriptorEmptyJSON,
+		Layers:       artifactManifestLayers,
+	}
+
+	rawData, err := json.Marshal(artifactManifest)
+	if err != nil {
+		return nil, err
+	}
+	if err := imageDest.PutManifest(ctx, rawData, nil); err != nil {
+		return nil, err
+	}
+	unparsed := newUnparsedArtifactImage(ir, artifactManifest)
```

Comment:
It would be correct for `unparsed.Manifest()` to return exactly `rawData`, without `unparsedArtifactImage.Manifest` building a new single-use structure.

AFAICS the current implementation will _probably_ return the same bytes as `rawData`, but with a mismatching `MediaTypeImageIndex` MIME type.

(Ultimately none of this matters _currently_ because `oci/layout` ignores the `unparsedTopLevel` parameter.)